src/dockerfile.ml

```diff
@@ -336,8 +338,24 @@ let string_of_run ~escape mounts network security c =
       (function `Insecure -> "insecure" | `Sandbox -> "sandbox")
       security
   in
+  mounts @ network @ security
+
+let string_of_run ~escape mounts network security c =
+  let params = string_of_run' ~escape mounts network security in
   let run = string_of_shell_or_exec ~escape c in
-  String.concat " " (mounts @ network @ security @ [ run ])
+  String.concat " " (params @ [ run ])
+
+let string_of_run_heredoc ~escape mounts network security c =
+  let params = string_of_run' ~escape mounts network security in
+  let cmds, docs =
+    List.fold_left
+      (fun (cmds, docs) (t, cmd) ->
+        let cmd = escape_string ~char_to_escape:'\n' ~escape cmd in
+        ( cmds @ [ sprintf "<<%s%s %s" (if t.strip then "-" else "") t.word cmd ],
```

Comment:
there should be a special case for when `cmd` is empty. e.g. the example from the reference documentation:
```
# syntax=docker/dockerfile:1
FROM debian
RUN <<EOT
  mkdir -p foo/bar
EOT
```